### PR TITLE
Tiles are now (mostly) pooled objects

### PR DESCRIPTION
**Pull request by coiax**
Floors no longer have a builtin_tile, but instead use PoolOrNew(). Mostly. Some things still use new, but turfs use pooled tiles.

Also added a do-nothing SSpool so you can inspect the global pool.

### CODE REVIEW

**Comment by xxalpha** [comment]
Does this actually improve speed?

**Comment by coiax** [comment]
It should. I discussed this change with MSO, tiles are destroyed constantly during a singuloth escape, while it uses the pregenerated tiles when it's ripped out of the floor. Making the tiles pooled should help our startup times by saving on an object per open turf.

**Comment by optimumtact** [comment]
Just aiming to get a test merge and a few singulo releases before I pull the trigger on this.

**Comment by Fox-McCloud** [comment]
This won't help performance because pooling has a net loss compared to garbage collecting it out. Creation of the object is a hair tinyyy bit cheaper (not much, and certainly barely worth noting), but pooling the object is far, far more expensive than just letting it garbage collect.

Pooling is nice, in theory, but once you profile it, it becomes apparent how detrimental it is.

**Comment by coiax** [comment]
@RemieRichards ^ ?

**Comment by RemieRichards** [comment]
I don't recall us ever finding anything to corroborate Fox's claims -for tg-
you have to remember we have VERY beefy server hardware, and DM is known to act wildly different based on the stats of the host machine.

**Comment by optimumtact** [comment]
Nobody ever proved pooling was definitively better either

**Comment by xxalpha** [comment]
>I don't recall us ever finding anything to corroborate Fox's claims -for tg-

Except we have.

https://github.com/tgstation/tgstation/pull/13279

Now it's better, but does it perform better in this case? You should profile it.

Is creating or destroying a floor tile that expensive that you need to add the pooling overhead?

**Comment by coiax** [comment]
I thought the whole point was that pooling wasn't an overhead, and that it was cheaper than creating and destroying the tiles No measurable difference in startup times between mainline and moving to pooled tiles.

**Comment by coiax** [comment]
@MrStonedOne would be interested in your thoughts on this.

**Comment by MrStonedOne** [comment]
Part of this has to do with the server start up overhead of creating the 22 thousand tiles, and the memory usage they bring. further more, under byond 511, new() cost gets higher, and var read/write cost gets lower, so I'd still like to move forward with this.

**Comment by Fox-McCloud** [comment]
@xxalpha 

That change didn't make it viable, it just made it better; it still doesn't cost less, overall, to pool an item, than to let it GC out.

If you pool it, pulling it out of the pool to `new` it is cheaper, *barely*, but the cost to reset vars is still incredibly high compared to plain old garbage collection.

**Comment by coiax** [comment]
I did a test of how many roundstart tiles are generated. There are approximately 13,000 tile objects generated on Box, and 5000 more from the random templates I got with this test.

**Comment by Jalleo** [comment]
How about having it create the pool at the start of the round and slowly it gives each turf its own tile and remove it off the pool? This should mean it wont be too laggy if we are using new in spare ticks.

**Comment by xxalpha** [comment]
Might as well use SS initialization, it is not a good idea to set things up during a round.

**Comment by coiax** [comment]
I did propose initially the idea of a "lazy process" which is a processing list that runs on spare ticks, but MSO suggested pooling tiles because they create and destroy each other all the time when they cross over each other.

**Comment by MrStonedOne** [comment]
>not rebasing

**Comment by coiax** [comment]
Whenever a PR gets messy and has more than one clean commit, I just merge, rather than have to go through merge conflicts in half commits that I probably commited because Iw as going to bed or doing something else

File: code/controllers/subsystem/pool.dm [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	flags = SS_BACKGROUND | SS_FIRE_IN_LOBBY
+	var/list/global_pool
+	var/list/pool_levels = list()
+	var/sum = 0
+
+	var/list/maintained_types = list(
+		/obj/item/stack/tile/plasteel = 100
+	)
+
+	var/stats_placed_in_pool = list()
```
Comment (RemieRichards) [inline code/controllers/subsystem/pool.dm]: type your lists ```/list```

File: code/controllers/subsystem/pool.dm [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	else
+		msg += "NULL POOL"
+	..(msg)
+
+/datum/subsystem/pool/fire()
+	sum = 0
+	for(var/type in global_pool + maintained_types)
+		var/list/L = global_pool[type]
+		var/required_number = 0
+		if(type in maintained_types)
```
Comment (RemieRichards) [inline code/controllers/subsystem/pool.dm]: ```if(maintained_types[type])``` is faster

File: code/__DEFINES/misc.dm [hunk 3]
```diff
@@ -474,3 +474,4 @@ var/global/list/ghost_others_options = list(GHOST_OTHERS_SIMPLE, GHOST_OTHERS_DE
 #define debug_world_log(msg) if (Debug2) world.log << "DEBUG: [msg]"
 
 #define COORD(A) "([A.x],[A.y],[A.z])"
+#define INCREMENT_TALLY(L, stat) if(stat in L){L[stat]++}else{L[stat] = 1}
```
Comment (RemieRichards) [inline code/__DEFINES/misc.dm]: ```if(L[stat])```

Comment (xxalpha) [inline code/__DEFINES/misc.dm]: Lists always have associations?

Comment (RemieRichards) [inline code/__DEFINES/misc.dm]: @xxalpha ones using this macro will be, as you can see from the ```L[stat] = 1``` at the end.

**Comment by ChangelingRain** [comment]
pooling is probably a good thing :tm:

**Comment by Fox-McCloud** [comment]
>pooling is probably a good thing

Which would be worth saying if you actually bothered profiling it.

**Pull request merged**